src/opentimelineio/algo/editAlgorithm.h

```diff
@@ -0,0 +1,188 @@
+
+#pragma once
+
+#include "opentimelineio/composition.h"
+
+namespace opentimelineio { namespace OPENTIMELINEIO_VERSION { namespace algo {
+
+//! Enum used by 3/4 Point Edit (aka. as fill)
+enum class ReferencePoint
+{
+    Source,
+    Sequence,
+    Fit
+};
+
+// Overwrite an item or items.
+//
+// | A | B |  ->  |A| C |B|
+//   ^   ^
+//   | C |
+//
+//          item = item to overwrite (usually a clip)
+//   composition = usually a track item.
+//         range = time range to overwrite.
+// remove_transitions = whether to remove transitions within range.
+// fill_template = item to fill in (usually a gap),
+//                 when range > composition's time.
+void overwrite(
+    Item*            item,
+    Composition*     composition,
+    TimeRange const& range,
+    bool             remove_transitions = true,
+    Item*            fill_template      = nullptr,
+    ErrorStatus*     error_status       = nullptr);
+
+// Insert an item.
+// |     A     | B |  ->  | A | C | A | B |
+//       ^
+//     | C |
+//
+//          item = item to insert (usually a clip)
+//   composition = usually a track item.
+//          time = time to insert at.  If < composition's start time, it will insert at 0 index.
+//                                     If > composition's end_time_exclusive, it will append at end.
+// remove_transitions = whether to remove transitions that intersect time.
+// fill_template = item to fill in (usually a gap),
+//                 when time > composition's time.
+void insert(
+    Item* const         item,
+    Composition*        composition,
+    RationalTime const& time,
+    bool const          remove_transitions = true,
+    Item*               fill_template      = nullptr,
+    ErrorStatus*        error_status       = nullptr);
+
+//
+// Adjust a single item's start time or duration.
+//
+// |    A    | B | C |  ->  |  A  |FILL| B | C |
+//        <--*
+//            
+//          item = Item to apply trim to (usually a clip)
+//      delta_in = RationalTime that the item's source_range().start_time()
+//                 will be adjusted by
+//     delta_out = RationalTime that the item's
+//                 source_range().end_time_exclusive() will be adjusted by
+// fill_template = item to fill in (usually a gap),
+//                 when time > composition's time.
+void trim(
+    Item*               item,
+    RationalTime const& delta_in,
+    RationalTime const& delta_out,
+    Item*               fill_template = nullptr,
+    ErrorStatus*        error_status  = nullptr);
+
+// Slice an item.
+//
+// | A | B | -> |A|A| B |
+//   ^
+// composition = usually a track item.
+//        time = time to slice at.
+void slice(
+    Composition*        composition,
+    RationalTime const& time,
+    bool const          remove_transitions = true,
+    ErrorStatus*        error_status       = nullptr);
+
+//
+// Slip an item start_time by + or -, clamping to available_range if available.
+//
+// |   A   |
+//  <----->
+//
```

Comment:
how is the slip accomplished? what happens if the delta results in negative values?

Reply:
They are accepted as long as there's no available range or the available range supports the negative values.